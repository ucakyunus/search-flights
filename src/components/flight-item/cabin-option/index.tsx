import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { Box, Radio, Button, RadioProps, Typography, ButtonProps, useTheme } from '@mui/material';
import styled from '@emotion/styled';

export interface CabinOptionProps extends ButtonProps {
  price: number;
  label: string;
  currency: string;
  radioProps?: RadioProps;
  isSelected?: boolean;
}

export const CabinOption = ({
  price,
  label,
  currency,
  radioProps,
  isSelected = false,
  ...rest
}: CabinOptionProps) => {
  const { custom } = useTheme();

  const StyledButton = styled(Button)({
    boxShadow: custom?.shadows![1],
    width: 200,
    justifyContent: 'flex-start',
    backgroundColor: 'white',
    ...(isSelected && {
      ':after': {
        content: '""',
        position: 'absolute',
        bottom: -16,
        left: 0,
        width: '100%',
        height: 16,
        backgroundColor: 'white',
      },
      ':hover': {
        backgroundColor: 'white',
      },
    }),
  });

  return (
    <StyledButton {...rest}>
      <Radio {...radioProps} />

      <Typography
        ml={-0.5}
        color='gray'
        fontSize={8}
        fontWeight={600}
        sx={{ textDecoration: 'underline' }}
      >
        {label}
      </Typography>

      <Box display='flex' flexDirection='column' alignItems='start' ml={1}>
        <Typography fontSize={8} color='gray'>
          Yolcu Basina
        </Typography>
        <Typography fontSize={14} fontWeight={600}>
          {currency} {price}
        </Typography>
      </Box>

      <ExpandMoreIcon sx={{ marginLeft: 'auto' }} />
    </StyledButton>
  );
};
