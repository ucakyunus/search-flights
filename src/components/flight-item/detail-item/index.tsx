import { useNavigate } from 'react-router-dom';
import { Box, Button, Typography } from '@mui/material';

import { BrandCode, ResultTypes, Status } from '@/constants';

import type { FareSubcategory } from '@/types/flight';

export interface DetailItemProps {
  detail: FareSubcategory;
  isPromotionActive: boolean;
}

export const DetailItem = ({ detail, isPromotionActive }: DetailItemProps) => {
  const { status, price, brandCode, rights } = detail;

  const navigate = useNavigate();

  const isAvailable = status === Status.AVAILABLE;
  const isEcoFly = brandCode === BrandCode.ECO_FLY;

  const isButtonDisabled = isPromotionActive && !isEcoFly;

  const newPrice = isEcoFly && isPromotionActive ? price.amount / 2 : price.amount;

  const onSelectHandler = () => {
    const state = isAvailable
      ? {
          status: ResultTypes.Success,
          amount: newPrice,
          currency: price.currency,
        }
      : { status: ResultTypes.Failure };

    navigate('/result', { state });
  };

  return (
    <Box
      width={400}
      display='flex'
      minHeight={300}
      alignItems='stretch'
      flexDirection='column'
      sx={{ backgroundColor: 'white' }}
    >
      <Box
        p={1}
        height={60}
        display='flex'
        alignItems='flex-start'
        bgcolor='secondary.light'
        justifyContent='space-between'
      >
        <Typography fontSize={16} fontWeight={600} textTransform='capitalize'>
          {brandCode}
        </Typography>

        <Box display={'flex'} gap={1}>
          {isPromotionActive && isEcoFly && (
            <Box
              display='flex'
              alignItems='flex-start'
              sx={{ textDecoration: 'line-through', opacity: 0.4 }}
            >
              <Typography fontSize={12}>{price.currency}</Typography>
              <Typography fontSize={16} ml={0.5}>
                {price.amount}
              </Typography>
            </Box>
          )}
          <Box display='flex' alignItems='flex-start'>
            <Typography fontSize={12}>{price.currency} </Typography>
            <Typography fontSize={16} ml={0.5}>
              {newPrice}
            </Typography>
          </Box>
        </Box>
      </Box>
      <Box flex={1} border={1} borderColor='secondary.light'>
        {rights.map((right, index) => (
          <Typography
            p={1}
            key={index}
            fontSize={12}
            borderBottom={1}
            borderColor='secondary.light'
          >
            {right}
          </Typography>
        ))}
      </Box>
      <Button
        fullWidth
        size='large'
        color='error'
        variant='contained'
        onClick={onSelectHandler}
        disabled={isButtonDisabled}
      >
        Uçuşu Seç
      </Button>
    </Box>
  );
};
