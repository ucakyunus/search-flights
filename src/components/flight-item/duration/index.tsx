import { Box, BoxProps, Typography } from '@mui/material';

export interface DurationProps extends BoxProps {
  time: string;
}

export const Duration = ({ time, ...rest }: DurationProps) => {
  return (
    <Box
      display='flex'
      flexDirection='column'
      alignItems='center'
      justifyContent={'center'}
      {...rest}
    >
      <Typography color='gray' noWrap>
        Uçuş Süresi
      </Typography>
      <Typography fontWeight={600}>{time}</Typography>
    </Box>
  );
};
