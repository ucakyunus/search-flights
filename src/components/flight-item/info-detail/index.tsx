import { Box, BoxProps, Typography } from '@mui/material';

export interface InfoDetailProps extends BoxProps {
  time: string;
  portCode: string;
  cityName: string;
}

export const InfoDetail = ({ time, portCode, cityName, ...rest }: InfoDetailProps) => {
  return (
    <Box {...rest}>
      <Typography fontSize={14} fontWeight={600}>
        {time}
      </Typography>
      <Typography fontSize={14} color='gray'>
        {portCode}
      </Typography>
      <Typography fontSize={10} color='gray'>
        {cityName}
      </Typography>
    </Box>
  );
};
