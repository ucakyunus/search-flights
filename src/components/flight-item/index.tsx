import { useMemo } from 'react';
import { Box, Divider, useTheme } from '@mui/material';

import { CabinOption } from '@/components/flight-item/cabin-option';
import { InfoDetail } from '@/components/flight-item/info-detail';
import { Duration } from '@/components/flight-item/duration';
import { DetailItem } from '@/components/flight-item/detail-item';
import { FareCategories } from '@/constants';

import type { Flight } from '@/types/flight';

export interface FlightItemProps {
  index: number;
  data: Flight;
  isPromotionActive: boolean;
  selectedFareOption: string | null;
  setSelectedFareOption: (value: string) => void;
}

export const FlightItem = (props: FlightItemProps) => {
  const { custom } = useTheme();

  const { data, index, isPromotionActive, selectedFareOption, setSelectedFareOption } = props;

  const { fareCategories } = data;

  const businessRadioValue = `${FareCategories.BUSINESS}-${index}`;
  const economyRadioValue = `${FareCategories.ECONOMY}-${index}`;
  const isBusinessSelected = selectedFareOption === businessRadioValue;
  const isEconomySelected = selectedFareOption === economyRadioValue;

  const flightItemDetailItems = useMemo(() => {
    if (!selectedFareOption) return [];

    if (selectedFareOption !== businessRadioValue && selectedFareOption !== economyRadioValue)
      return [];

    const [selectedCabinType] = selectedFareOption.split('-');

    const subcategories =
      selectedCabinType === FareCategories.BUSINESS
        ? fareCategories.BUSINESS.subcategories
        : fareCategories.ECONOMY.subcategories;

    return subcategories.sort((a, b) => a.order - b.order);
  }, [
    businessRadioValue,
    economyRadioValue,
    fareCategories.BUSINESS.subcategories,
    fareCategories.ECONOMY.subcategories,
    selectedFareOption,
  ]);

  const businessEcoFlyPrice = useMemo(
    () =>
      fareCategories.BUSINESS.subcategories.find((category) => {
        return category.order === 1;
      })?.price.amount as number,
    [fareCategories.BUSINESS.subcategories],
  );

  const economyEcoFlyPrice = useMemo(
    () =>
      fareCategories.ECONOMY.subcategories.find((category) => {
        return category.order === 1;
      })?.price.amount as number,
    [fareCategories.ECONOMY.subcategories],
  );

  const currency = useMemo(
    () => fareCategories.BUSINESS.subcategories[0].price.currency,
    [fareCategories.BUSINESS.subcategories],
  );

  const economyPriceShown = useMemo(() => {
    return isPromotionActive ? economyEcoFlyPrice / 2 : economyEcoFlyPrice;
  }, [economyEcoFlyPrice, isPromotionActive]);

  const businessPriceShown = useMemo(() => {
    return isPromotionActive ? businessEcoFlyPrice / 2 : businessEcoFlyPrice;
  }, [businessEcoFlyPrice, isPromotionActive]);

  return (
    <Box>
      <Box display='flex' gap={2}>
        <Box
          p={2}
          flex={1}
          display='flex'
          alignItems='center'
          justifyContent='space-between'
          boxShadow={custom?.shadows![1]}
          sx={{ backgroundColor: 'white' }}
        >
          <Box display='flex' alignItems='center' flex={1}>
            <InfoDetail
              time={data.departureDateTimeDisplay}
              portCode={data.originAirport.code}
              cityName={data.originAirport.city.name}
            />

            <Divider
              sx={{
                width: '400px',
                height: '1px',
                backgroundColor: '#000',
                mx: 1,
              }}
            />

            <InfoDetail
              textAlign='right'
              time={data.arrivalDateTimeDisplay}
              portCode={data.destinationAirport.code}
              cityName={data.destinationAirport.city.name}
            />
          </Box>

          <Duration mr={3} time={data.flightDuration} />
        </Box>

        <CabinOption
          label='Economy'
          currency={currency}
          price={economyPriceShown}
          isSelected={isEconomySelected}
          onClick={() => setSelectedFareOption(economyRadioValue)}
          radioProps={{
            value: economyRadioValue,
            checked: isEconomySelected,
            onChange: () => setSelectedFareOption(economyRadioValue),
          }}
        />

        <CabinOption
          label='Business'
          currency={currency}
          price={businessPriceShown}
          isSelected={isBusinessSelected}
          onClick={() => setSelectedFareOption(businessRadioValue)}
          radioProps={{
            value: businessRadioValue,
            checked: isBusinessSelected,
            onChange: () => setSelectedFareOption(businessRadioValue),
          }}
        />
      </Box>

      {flightItemDetailItems.length > 0 ? (
        <Box display='flex' gap={2} p={2} mt={2} bgcolor={'white'}>
          {flightItemDetailItems.map((item, index) => (
            <DetailItem key={index} detail={item} isPromotionActive={isPromotionActive} />
          ))}
        </Box>
      ) : null}
    </Box>
  );
};
