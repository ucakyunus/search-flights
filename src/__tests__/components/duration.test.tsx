import { render } from '@testing-library/react';
import { Duration, DurationProps } from '@/components/flight-item/duration';

const mockProps: DurationProps = {
  time: '2h 30m',
  mr: 3,
};

test('Duration matches snapshot', () => {
  const { asFragment } = render(<Duration {...mockProps} />);
  expect(asFragment()).toMatchSnapshot();
});
