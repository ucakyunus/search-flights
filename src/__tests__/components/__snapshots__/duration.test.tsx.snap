// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Duration matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1n8brx5"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-o6dx9f-MuiTypography-root"
    >
      Uçuş Süresi
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-izap9d-MuiTypography-root"
    >
      2h 30m
    </p>
  </div>
</DocumentFragment>
`;
